File: android/src/com/bombbird/terminalcontrol/PlayGamesManager.java
package com.bombbird.terminalcontrol;

import android.app.Activity;
import android.content.Intent;
import android.util.Log;
import android.widget.Toast;
import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.signin.*;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;

import static com.google.android.gms.common.ConnectionResult.SIGN_IN_REQUIRED;

public class PlayGamesManager {
    private static int RC_SIGN_IN = 777;

    protected void onActivityResult(Activity activity, int requestCode, int resultCode, Intent data) {
        if (requestCode == RC_SIGN_IN) {
            GoogleSignInResult result = Auth.GoogleSignInApi.getSignInResultFromIntent(data);
            if (result.isSuccess()) {
                // The signed in account is stored in the result.
                GoogleSignInAccount signedInAccount = result.getSignInAccount();
                Toast toast = Toast.makeText(activity, "Login successful", Toast.LENGTH_LONG);
                toast.show();
            } else {
                String message = result.getStatus().getStatusMessage();
                Log.e("e", String.valueOf(result.getStatus()));
                if (result.getStatus().getStatusCode() == SIGN_IN_REQUIRED) {
                    activity.startActivityForResult(GoogleSignIn.getClient(activity, GoogleSignInOptions.DEFAULT_GAMES_SIGN_IN).getSignInIntent(), RC_SIGN_IN);
                }
                /*
                if (message == null || message.isEmpty()) {
                    Toast toast = Toast.makeText(activity, R.string.signin_other_error, Toast.LENGTH_LONG);
                    toast.show();
                } else {
                    Toast toast = Toast.makeText(activity, message, Toast.LENGTH_LONG);
                    toast.show();
                }
                */
            }
        }
    }

    protected void gameSignIn(Activity activity) {
        if (GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(activity) != ConnectionResult.SUCCESS) {
            Toast toast = Toast.makeText(activity, "Google Play Games is unavailable; please check your installation on Google Play Store.", Toast.LENGTH_LONG);
            toast.show();
        }
        GoogleSignInOptions signInOptions = GoogleSignInOptions.DEFAULT_GAMES_SIGN_IN;
        GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(activity);
        if (GoogleSignIn.hasPermissions(account, signInOptions.getScopeArray())) {
            // Already signed in.
            // The signed in account is stored in the 'account' variable.
            GoogleSignInAccount signedInAccount = account;
        } else {
            // Haven't been signed-in before. Try the silent sign-in first.
            GoogleSignInClient signInClient = GoogleSignIn.getClient(activity, signInOptions);
            signInClient.silentSignIn().addOnCompleteListener(
                activity, task -> {
                    if (task.isSuccessful()) {
                        // The signed in account is stored in the task's result.
                        GoogleSignInAccount signedInAccount = task.getResult();
                    } else {
                        // Player will need to sign-in explicitly using via UI.
                        // See [sign-in best practices](http://developers.google.com/games/services/checklist) for guidance on how and when to implement Interactive Sign-in,
                        // and [Performing Interactive Sign-in](http://developers.google.com/games/services/android/signin#performing_interactive_sign-in) for details on how to implement
                        // Interactive Sign-in.
                        startSignInIntent(activity);
                    }
                });
        }
    }

    protected void startSignInIntent(Activity activity) {
        GoogleSignInClient signInClient = GoogleSignIn.getClient(activity, GoogleSignInOptions.DEFAULT_GAMES_SIGN_IN);
        Intent intent = signInClient.getSignInIntent();
        activity.startActivityForResult(intent, RC_SIGN_IN);
    }
}

File: android/src/com/bombbird/terminalcontrol/PlayGamesManager.kt
package com.bombbird.terminalcontrol

import android.view.Gravity
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import android.widget.Toast
import com.bombbird.terminalcontrol.entities.achievements.UnlockManager
import com.bombbird.terminalcontrol.utilities.PlayGamesInterface
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.games.Games
import com.google.android.gms.tasks.Task

class PlayGamesManager(private val activity: AndroidLauncher): PlayGamesInterface {
    var signedInAccount: GoogleSignInAccount? = null
        set(value) {
            if (field == null && value != null) {
                val client = Games.getGamesClient(activity, value)
                client.setViewForPopups(activity.view)
                client.setGravityForPopups(Gravity.TOP or Gravity.CENTER_HORIZONTAL)
            }
            field = value
            if (value != null && UnlockManager.achievementList.size > 0) UnlockManager.checkGooglePlayAchievements()
        }

    override fun gameSignIn() {
        if (GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(activity) != ConnectionResult.SUCCESS) {
            activity.runOnUiThread {
                val toast = Toast.makeText(
                    activity,
                    "Google Play Games is unavailable; please check your installation on Google Play Store.",
                    Toast.LENGTH_LONG
                )
                toast.show()
            }
        }
        val signInOptions = GoogleSignInOptions.DEFAULT_GAMES_SIGN_IN
        val account = GoogleSignIn.getLastSignedInAccount(activity)
        if (GoogleSignIn.hasPermissions(account, *signInOptions.scopeArray)) {
            //Already signed in
            //The signed in account is stored in the 'account' variable
            signedInAccount = account
        } else {
            //Haven't been signed-in before; try the silent sign-in first
            val signInClient = GoogleSignIn.getClient(activity, signInOptions)
            signInClient.silentSignIn().addOnCompleteListener(
                activity
            ) { task: Task<GoogleSignInAccount?> ->
                if (task.isSuccessful) {
                    //The signed in account is stored in the task's result
                    signedInAccount = task.result
                } else {
                    //Player will need to sign-in explicitly using via UI
                    startSignInIntent()
                }
            }
        }
    }

    private fun startSignInIntent() {
        val signInClient = GoogleSignIn.getClient(activity, GoogleSignInOptions.DEFAULT_GAMES_SIGN_IN)
        val intent = signInClient.signInIntent
        activity.startActivityForResult(intent, AndroidLauncher.PLAY_SIGN_IN)
    }

    override fun gameSignOut() {
        val signInClient = GoogleSignIn.getClient(
            activity,
            GoogleSignInOptions.DEFAULT_GAMES_SIGN_IN
        )
        signInClient.signOut().addOnCompleteListener(activity) {
            //User is signed out
            signedInAccount = null
        }
    }

    override fun isSignedIn(): Boolean {
        return signedInAccount != null
    }

    override fun showAchievements() {
        signedInAccount?.let {
            Games.getAchievementsClient(activity, it).achievementsIntent.addOnSuccessListener { intent ->
                activity.startActivityForResult(intent, AndroidLauncher.PLAY_SHOW_ACHIEVEMENTS)
            }
        }
    }

    override fun unlockAchievement(id: String) {
        signedInAccount?.let {
            Games.getAchievementsClient(activity, it).unlock(id)
        }
    }

    override fun incrementAchievement(id: String, steps: Int, set: Boolean) {
        signedInAccount?.let {
            if (set) Games.getAchievementsClient(activity, it).setSteps(id, steps)
            else Games.getAchievementsClient(activity, it).increment(id, steps)
        }
    }
}